### ISSUE TITLE
report an error when assigning a default value to `self` or `cls`

### ISSUE THREAD

**Issue by DetachHead** [astral-sh/ruff#25820]
```py
class Foo:
    def foo(self=1): ...
```

this is almost certainly a mistake, so it should be reported as an error

maybe related: astral-sh/ty#3666, but i think this should always be an error, regardless of whether the default value is the correct type (maybe in that case it's better off as a ruff rule since it doesn't need type information?)

[basedpyright has a rule for this](https://docs.basedpyright.com/latest/benefits-over-pyright/new-diagnostic-rules/#reportselfclsdefault)

**Comment by carljm** [comment]
I think this should probably be a Ruff rule?

**Comment by assadyousuf** [comment]
@ntBre  Mind if i take this up? Is any discussion needed here? Im assuming we just report and error when self or cls is assigned a default value in a func signature

**Comment by MichaReiser** [comment]
This sounds reasonable to me. The analysis should not be limited to `self` and `cls` (by name), instead, it should flag on a method's first argument. 

The autofix can't be safe (if we even want one), because this pattern may be used to support different calling conventions. 

We need to make sure to correctly handle `@classmethod`, `__new__`, ..., `@override` (any decorators?)

I suggest taking a look at the basedpyright implementation, which has probably figured out most of the issues.

**Comment by RafaelJohn9** [comment]
> basedpyright has a rule for this

Basedpyright doesn't seem to cover the different cases:  https://github.com/DetachHead/basedpyright/blob/main/packages/pyright-internal/src/analyzer/checker.ts#L7677-L7692

**Comment by MichaReiser** [comment]
It handles some of them based on the call site

**Comment by RafaelJohn9** [comment]
@MichaReiser 

With BasedPyright: the check itself is intentionally very small and just looks at the first parameter, with the method classification happening through where it’s called.

For Ruff, I’m thinking of structuring it around an explicit “receiver slot” classification first, i.e. determine whether a direct class-body function is:

* instance-bound
* class-bound
* unbound

and only then check whether the receiver parameter (the first parameter for bound methods) has a default.

That would cover the same semantic space, but with the method classification made explicit in the rule logic itself rather than only in the call path. Concretely, the cases I’m planning for are:

* plain methods - instance-bound
* `@classmethod` - class-bound
* `__new__` - class-bound
* `@staticmethod` - unbound
* nested / non-method functions - unbound

This is what I had in mind with the implementation. Would this suffice or should we go with BasedPyright minimal route ?

**Comment by MichaReiser** [comment]
I'm not very familiar with ruff in this area but that sounds reasonable to me

**Comment by RafaelJohn9** [comment]
sure no worries 🤝 

I'll do this then
